### ISSUE TITLE
Provide deck shuffle option / added cards not showing in full random order

### ISSUE THREAD

**Issue by WantALongerUserName** [ankidroid/Anki-Android#5635]
Note: this is a duplicate of a reddit post I made here:
https://www.reddit.com/r/Anki/comments/dyahzy/daily_cards_are_not_fully_randomized_ankidroid/

###### Expected Result
- Daily cards in deck are fully randomized with no bias.

###### Actual Result
- Cards which were later added to the deck are always bias towards the end of the study session.

Summary of what I've seen:

I'm learning Chinese, and so year ago, I found and imported a deck of 5000 pre-made cards. They're wonderful, for each comes with an example sentence spoken by a native speaker. However, about 1000 cards in, I found that I disliked the sequence of the cards, as there were cards deeper in the deck that I wanted to learn sooner. I managed this by:

card browser => select card => reschedule card => 1 day

This worked great. The next day, those cards would be added to my daily deck. I then found that I wanted to learn new words that were not in the pre-made set, so I created some. And again, I rescheduled those new-created cards to be rescheduled for the next day. This also seemed to work great. Except I noticed something odd. No matter how old the newly created cards are (ie: their lapse time is now several weeks), they always appear near the end of my daily study deck. (An odd note: there's not a binary transition between the original and new cards, but some slight inter-mixing, as though a very incomplete shuffle was attempted. The newly-made cards are always in the last 10% of the daily deck.)

I've tried my best to search for an answer, and so far have been unable to. For the recorded, I have selected:
Settings => Reviewing => New card position: "mix new cards and reviews"
Deck => Options => New Cards => Order: "New cards in random order"
Settings => Preferences => Advanced => Experimental V2 Scheduler: not selected (would this help if I select?)

AnkiDroid 2.9.1

If easier, a possible alternative to tracking down this issue would be to provide a "randomize now" button such that at the beginning of a study session, the cards can be shuffled once more, hopefully into a truly random order. (I appreciate this could not be done once the session has started due to the time constraints on the queue for incorrectly answered cards.)

[ X ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ X ] I have searched for similar existing issues here and on the user forum

**Comment by HairyPhil** [comment]
When I reported this I was told that it was working as intended and that I should just pretend that reverse order is random order. Clearly whoever designed this process only uses finite decks that are constructed once and never have cards added to them, whereas users such as myself start with an empty deck and add cards daily.

**Comment by mikehardy** [comment]
Hmm - when you say "reported this", exactly where - anki.tenderapp.com upstream and you were told by user 'dae'/Damian Elmes? Or did you report it here and a maintainer here said that? Either way do you have a link to that so I could step through the reasoning?

In general we here / AnkiDroid stick diligently to upstream behavior for ecosystem compatibility so if that's what is indicated upstream we will follow it, but if it was just local to AnkiDroid maybe we can alter behavior

**Comment by HairyPhil** [comment]
https://anki.tenderapp.com/discussions/ankidesktop/32475-random-order-not-applied-to-new-cards

**Comment by mikehardy** [comment]
Thank you, I just read through that. It is true that the performance is heavily optimized for review speed as that is the primary use case (and by far the vast majority of time spent in the app, based on our stats). Combine that with the wide array of phones we support including some really really old ones (even e-readers) I can sympathize with the argument that shuffling is something that can't happen as a side-effect of a typical user interaction (review or add card). 

I can also sympathize with the desire though. The idea posed above of providing a "reshuffle" action available to the user is a possibility then. A bunch of other actions were added recently so there is recent work that could be used as a template for adding a new deck action if anyone was up to proposing a pull request for it

**Comment by WantALongerUserName** [comment]
I can't stress enough my appreciation of the efforts of those who contribute to this fantastic program. I use it everyday for studying Chinese. If a "shuffle deck option" could be provided on a future release, that would be invaluable.

**Comment by david-allison-1** [comment]
This has been in the back of my mind for a while.

It's due to the implicit ordering in the database.

Assume one deck with 100 cards. We currently take 50 cards, implicitly ordered by cid (timstamp of creation). We then shuffle these cards, dequeue one by one, and repeat when the queue is empty. All cards in the first 50 will be older than the cards in the second group of 50.

https://github.com/ankidroid/Anki-Android/blob/b3d6cb38f3412905462ca49c74ba8baaf66282b6/AnkiDroid/src/main/java/com/ichi2/libanki/Sched.java#L1254-L1268

This can be fixed by giving the user control over `Sched.mQueueLimit` (setting it to greater than the number of cards in the review queue means full randomisation).

Worst case for a user assumptions (40k cards needing review): 
* Long: 24 bytes (ptr + long + overhead)
* LinkedListNode<Long>: 56 bytes (ptr + Long + next + prev + overhead)
* 10,000 cards * 4 queues * LinkedListNode<Long>  = ~2.25MB

I've just fixed it on my phone. Going from 50 to ~700 cards has no noticeable impact in speed, and the reviews seem much more random (could be the placebo effect). I believe this would be over my full ~1300 cards on the V2 Scheduler.

I'd really like to push for an advanced setting to let the user set & disable this limit, I'll almost certainly keep it hacked in on my phone.

**Comment by mikehardy** [comment]
I could accept an advanced setting that turned on an advanced shuffle mode. I'd be much more keen on it if it was upstream in the desktop code first so that libanki didn't have non-upstream code - we work hard to maintain fidelity there. But if it was pretty surgical with how it sliced into libanki so that it was crystal clear which part was libanki and which part wasn't, and the default was upstream behavior I'd merge it

**Comment by david-allison-1** [comment]
Okay, happy to have this assigned to me then. 

Additional code in libAnki is going to be a setter in `Sched.java` and `SchedV2.java`

**Comment by val59000** [comment]
Hello @david-allison-1  and @mikehardy I understand the reasoning of both side of the discussion.
I'm wondering if the advanced option and true randomization described in the issue has been implemented yet ?
Is it available in the beta release ?
Thanks !

**Comment by david-allison-1** [comment]
@val59000 I couldn't think of a good way to explain what we're doing without a massive wall of text. I'll come back to it at some point.

Feel free to set up a fork and set `mQueueLimit` to 99999

https://github.com/ankidroid/Anki-Android/blob/71225e9e21f637bb5f7371c47b39cf70992570f6/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java#L152

**Comment by Arthur-Milchior** [comment]
> Assume one deck with 100 cards. We currently take 50 cards, implicitly ordered by cid (timstamp of creation). We then shuffle these cards, dequeue one by one, and repeat when the queue is empty

I fear you are misunderstanding how suffling works. When the deck option is switch to "random", all cards get a random due value. Then when new cards are added to the deck, they get the next value, greater than all current due. Ideally, the code should state "the card is added to a randomized deck, give a random due value". However, even this would be far from perfect because the due value does not change when a card's deck is changed. I believe the only correct way to deal with it is to ignore due entirely, that would really be preferable in my opinion. Anyway, in non randomized deck, due and nid are in the same order.

We could get random value from sql by doing `order by rand()`, but it would breaks to current behaviour:
* if you see a card C today, you'll see C's sibling tomorrow, so that you discover all cards of a note one after the other. This is done thanks to  the `due` value (which is theoretically shared by all cards of the same note, but this breaks when you randomize your deck or if you have cards in different deck...) 
* you want to select the first new card of a deck first. If you did `order by ord, rand()` you'll get all "Front" card first and you'll get "Back" card only after there is no more "front", which is not the goal here. But if you do `order by rand(), ord`, then "ord" will get ignored. The only way to do it would be
```
let r= rand();
order by hash(r+nid), ord
```
assuming sqlite admits variable (I believe it does) and a hash function. But hash are costly

**Comment by david-allison-1** [comment]
@Arthur-Milchior I was referring to

https://github.com/ankidroid/Anki-Android/blob/8b4df2ab90b9c7fbcd50216a191aafebb02b31a4/AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java#L735-L736

Which is fixed in SchedV2, but it's inefficient (n*log(n) on each query), grabbing all valid IDs (ideally with a better data structure than a linked list) would be a better time/memory tradeoff for performance

**Comment by idkw** [comment]
I m'a juste a regular user so I'll just write some feedback for you to consider.

In my opinion as a user:
- shuffling should remain as it is like you said (adding new cards after shuffling adds them in order after the max due number). 
- Shuffling all cards after every added card is not performant on some devices so it should be propose as an option that only desktop users or high end smartphones user would like to use. 
- Reshuffling with a single click button should be available easily because currently we need to use the desktop version to "reposition with random order" the resync.
- shuffling with or without burying siblings is not very practical. Usually you get random front cards the first few days then the same reverse cards the next few days. I understand this the current behavior but the due value being shared by all siblings is very inconvenient to me. When I shuffle I would expect to shuffle all cards including siblings indifferently. I don't want reverse card to be picked up independently from front cards. I want to shuffle everything. I've found an unanswered reddit question that describe exactly this issue : https://www.reddit.com/r/Anki/comments/7ve9rt/how_do_i_randomize_new_cards_including_reverse/?utm_medium=android_app&utm_source=share 


Why I need to shuffle everything indifferently ?

For some domains shuffling make no sense I imagine. But for my use case when studying language vocabulary I need to test my ability to quickly remember any word at any time. If the app show me only new front card or only new reverse card or both front and review at the same time I'm not training properly because it is too easy to "cheat" with short term memory

**Comment by Arthur-Milchior** [comment]
@idkw I entirely agree that shuffle is pretty bad. Actually, ordering is bad right now, it makes no sens at all. But, as you said, that's mostly a matter for upstream, as we try to stay close to anki in term of user experience

**Comment by idkw** [comment]
I have just found a feature I didn't noticed before. When long pressing a deck you can start a custom review session with randomized cards of the size of your choosing. I tried it and it worked perfectly. This is exactly what I expect of a randomized cards session. Could the same algorithm be promoted as a standard review option instead of having to create such a new custom review session ?
